javascript/ql/src/Security/CWE-078/IndirectCommandInjection.qhelp

```diff
@@ -0,0 +1,122 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+<overview>
+
+	<p>
+
+		Forwarding command-line arguments to
+		<code>require('child_process').exec</code>, or some other library
+		routine that executes a system command can change the meaning of the
+		command unexpectedly due to unescaped special characters.
+
+	</p>
+
+	<p>
+
+		When the forwarded command-line arguments have been typed by a
+		programmer in an interactive shell, an unexpected change is usually
+		discovered immediately because the command fails, a manual
+		escaping of any special characters fixes the issue in this case.
+
+	</p>
+
+	<p>
+
+		When the forwarded command-line arguments come from a parent
+		process, then the lack of escaping indirectly introduces a command
+		injection vulnerability in the parent process.
+
+	</p>
+
+</overview>
+<recommendation>
+
+	<p>
+
+		If possible, use hard-coded string literals to specify the
+		command to run or library to load. Instead of forwarding the
+		command-line arguments to the process, examine the command-line
+		arguments and then choose among hard-coded string literals.
+
+	</p>
+
+	<p>
+
+		If the applicable libraries or commands cannot be
+		determined at compile time, then add code to verify that the each forwarded
+		command-line argument is properly escaped before using it.
+
+	</p>
+
+	<p>
+
+		If the forwarded command-line arguments are part of the
+		arguments of the system command, prefer a library routine that handles
+		the arguments as an array of strings rather than a single string.
+
+	</p>
+
+</recommendation>
+<example>
+
+	<p>
+
+		The following example shows code that executes another
+		JavaScript file in a child process by forwarding the command-line
+		arguments that a user has provided. This is problematic because the
+		special characters in the input from the user may change the meaning
+		of the child process unexpectedly. For instance, if one of the
+		command-line arguments is <code>"dollar$separated$name"</code>, then
+		the new process will substitute the two shell variables
+		<code>$separated</code> and <code>$name</code> before invoking
+		<code>node</code>.
+
+	</p>
+
+	<sample src="examples/indirect-command-injection.js" />
+
+	<p>
+
+		To guard against this, use a function, such as
+		<code>child_process.execFile</code>, that does not make the shell
+		perform substitution:
+
+	</p>
+
+	<sample src="examples/indirect-command-injection_fixed.js" />
+
+	<p>
+
+		Alternatively, the programmer needs to prevent substitutions
+		of the command-line arguments manually, for example by typing
+		<code>"dollar\$separated\$name"</code> or
+		<code>"'dollar$separated$name'"</code>.
```

Comment:
Is this something we want to suggest? Also, it's not clear who "the programmer" is.